src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,551 @@
+                {
+                    "file": {"path": file_path},
+                    "workspace_name": workspace_name,
+                    "assembly_name": assembly_name,
+                }
+            )
+            success = True
+        except Exception as e:
+            print(f"Error:\n{e}\nfrom attempt {attempts + 1}.\nTrying to rerun.")
+            attempts += 1
+
+    if not success:
+        raise ValueError(
+            f"Upload Failed for {file_path} after {max_attempts} attempts!"
+        )
+
+
+def _read_yaml_file(
+        root_dir: str,
+        env: str,
+        workspace_id: int,
+        assembly_dir_name: str,
+        assembly_name: str, 
+) -> Tuple[dict[str, dict[int, list[str]]], bool]:
+    """
+    Get metadata and upload status of an assembly from the uploaded.yaml file.
+    """
+
+    uploaded = False
+    if env not in loader_common_names.KB_ENV:
+        raise ValueError(f"Currently only support these {loader_common_names.KB_ENV} envs for upload")
+
+    file_path = _get_yaml_file_path(root_dir, assembly_dir_name)
+
+    with open(file_path, "r") as file:
+        data = yaml.safe_load(file)
+
+    if not data:
+        data = {env: dict()}
+
+    if workspace_id not in data[env]:
+        data[env][workspace_id] = list()
+
+    assembly_dicts = data[env][workspace_id]
+    if assembly_name in [assembly_dict["file_name"] for assembly_dict in assembly_dicts]:
+        uploaded = True
+    return data, uploaded
+
+
+def _update_yaml_file(
+        root_dir: str,
+        env: str,
+        workspace_id: int,
+        upa: str,
+        assembly_dir_name: str,
+        assembly_name: str,
+) -> None:
+    """
+    Update the uploaded.yaml file in target genome_dir with newly uploaded assembly names and upa info.
+    """
+    data, uploaded = _read_yaml_file(root_dir, env, workspace_id, assembly_dir_name, assembly_name)
+
+    if uploaded:
+        raise ValueError(f"Assembly {assembly_name} already exists in workspace {workspace_id}")
+
+    data_dict = {"file_name": assembly_name, "upa": upa}
+    data[env][workspace_id].append(data_dict)
+
+    file_path = _get_yaml_file_path(root_dir, assembly_dir_name)
+    with open(file_path, "w") as file:
+        yaml.dump(data, file)
+
+
+def _fetch_assemblies_to_upload(
+        root_dir: str,
+        env: str,
+        workspace_id: int,
+        collection_source_dir: str,
+        upload_file_ext: list[str],
+) -> Tuple[int, dict[str, str]]:
+    """
+    Help fetch assemblies to upload.
+    """
+    count = 0
+    wait_to_upload_assemblies = dict()
+
+    assembly_dirs = [
+        os.path.join(collection_source_dir, d)
+        for d in os.listdir(collection_source_dir)
+        if os.path.isdir(os.path.join(collection_source_dir, d))
+    ]
+
+    for assembly_dir in assembly_dirs:
+        assembly_files = [
+            f
+            for f in os.listdir(assembly_dir)
+            if os.path.isfile(os.path.join(assembly_dir, f))
+        ]
```

Comment:
There should only ever be one assembly file per directory, right? I think the way I would do this is get the list of files that match the extension and then throw an error if that's != 1. Then check the yaml file

Reply:
Please correct me if I am wrong.

My assumption was it could have more than one assembly files depending on the extension given. I remember you once mentioned we may include more extensions other than just "genomic.fna.gz".

Reply:
We might have a genome file *and* and assembly file, but not multiple assembly files. 
Genome files would have a different extension, like `.gbk` or `.gff`
@Tianhao-Gu can you correct me if I'm wrong here?

Reply:
That's correct. But genome files are compressed with extension like `genomic.gbff.gz` and `genomic.gff.gz`.
here is an example genome link in NCBI:
https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/172/955/GCA_000172955.1_ASM17295v1/

Reply:
Updated!

Reply:
:+1: